wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WorkspaceUpgrader.java

```diff
@@ -0,0 +1,140 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Upgrader for workspaces. This upgrade should be run after the {@link ConnectionUpgrader}
+ */
+public class WorkspaceUpgrader {
+  private static final Logger LOG = LoggerFactory.getLogger(WorkspaceUpgrader.class);
+
+  private final UpgradeStore upgradeStore;
+  private final SystemHttpServiceContext context;
+  private final long upgradeBeforeTsSecs;
+  private final ConnectionDiscoverer discoverer;
+  private final WorkspaceStore wsStore;
+
+  public WorkspaceUpgrader(UpgradeStore upgradeStore, SystemHttpServiceContext context, long upgradeBeforeTsSecs,
+                           WorkspaceStore wsStore) {
+    this.upgradeStore = upgradeStore;
+    this.context = context;
+    this.upgradeBeforeTsSecs = upgradeBeforeTsSecs;
+    this.discoverer = new ConnectionDiscoverer(context);
+    this.wsStore = wsStore;
+  }
+
+  public void upgradeWorkspaces() throws Exception {
+    List<NamespaceSummary> namespaces = context.listNamespaces();
+    for (NamespaceSummary ns : namespaces) {
+      if (!upgradeStore.isWorkspaceUpgradeComplete(ns)) {
+        upgradeWorkspacesInConnections(ns);
+      }
+    }
+    upgradeStore.setWorkspaceUpgradeComplete();
+  }
+
+  private void upgradeWorkspacesInConnections(NamespaceSummary namespace) {
+    List<Workspace> workspaces = TransactionRunners.run(context, ctx -> {
+      WorkspaceDataset wsDataset = WorkspaceDataset.get(ctx);
+      return wsDataset.listWorkspaces(namespace, upgradeBeforeTsSecs);
+    });
+
+    for (Workspace workspace : workspaces) {
+      List<Row> sample = Collections.emptyList();
+      try {
+        sample = DirectivesHandler.fromWorkspace(workspace);
+      } catch (Exception e) {
+        // this should not happen, but guard here to avoid failing the entire upgrade process
+        LOG.warn("Error retrieving the sample for workspace {}", workspace.getName(), e);
+      }
+      List<String> directives = workspace.getRequest() == null ? Collections.emptyList() :
+                                  workspace.getRequest().getRecipe().getDirectives();
+      WorkspaceId workspaceId = new WorkspaceId(namespace, workspace.getNamespacedId().getId());
+
+      long now = System.currentTimeMillis();
+      io.cdap.wrangler.proto.workspace.v2.Workspace.Builder ws =
+        io.cdap.wrangler.proto.workspace.v2.Workspace
+          .builder(workspace.getName(), workspace.getNamespacedId().getId()).setDirectives(directives)
+          .setCreatedTimeMillis(now).setUpdatedTimeMillis(now);
+
+      String connectorName = SpecificationUpgradeUtils.getConnectorName(
+        workspace.getProperties().get(PropertyIds.CONNECTION_TYPE).toLowerCase());
+      String path = SpecificationUpgradeUtils.getPath(workspace);
+      // if the connection type cannot be upgraded, create a workspace with empty sample spec, so there will be
```

Comment:
rather than use null for this special purpose, it's cleaner to have a method that specifically checks if a given connection type can be upgraded.

Reply:
is this just for ADLS? Should we just skip upgrade for these types of workspaces since you won't be able to see them in the new UX anyway?

Reply:
it is not just for ADLS, also for workspaces created with direct upload, I think we will not do upgrade for ADLS but still do upgrade for the upload file, as the UX will support upload file to create a workspace

Reply:
Updated logic: now it will only upgrade workspaces in the allowed connection type list. Only UPLOAD type workspace will get created without a sample spec